language-plutus-core/src/Language/PlutusCore/Evaluation/Machine/Cek.hs

```diff
@@ -55,48 +124,112 @@ type VarEnv = UniqueMap TermUnique Closure
 data CekEnv = CekEnv
     { _cekEnvMeans  :: DynamicBuiltinNameMeanings
     , _cekEnvVarEnv :: VarEnv
+    , _cekEnvProtocolParameters :: ()
     }
 
+data ExBudget = ExBudget { _exBudgetCPU :: ExCPU, _exBudgetMemory :: ExMemory }
+
+-- TODO memory should probably be fixed size - but then how do we handle overflow?
+newtype ExMemory = ExMemory Integer -- Probably machine word size
+  deriving (Eq, Ord)
+  deriving newtype Num
+  deriving (Semigroup, Monoid) via (Sum Integer)
+newtype ExCPU = ExCPU Integer
+  deriving (Eq, Ord)
+  deriving newtype Num
+
+makeLenses ''ExBudget
+
 -- | The monad the CEK machine runs in.
-type CekM = ReaderT CekEnv (Either CekMachineException)
+type CekMBase w s = ExceptT CekMachineException (RWS CekEnv w s)
+type SpendExUnits w s = (SpendExUnitsState s, SpendExUnitsWriter w)
+type CekM w s a = SpendExUnits w s => CekMBase w s a
+
+data ExTally = ExTally
+    { _exTallyCPU :: MonoidalHashMap (Plain Term) [ExCPU]
+    , _exTallyMemory :: MonoidalHashMap (Plain Term) [ExMemory]
+    }
+    deriving stock (Eq, Generic)
+    deriving (Semigroup, Monoid) via (GenericSemigroupMonoid ExTally)
+
+spendBoth :: WithMemory Term -> ExCPU -> ExMemory -> CekM w s ()
+spendBoth term cpu mem = spendCPU term cpu >> spendMemory term mem
+
+spendMemory :: WithMemory Term -> ExMemory -> CekM w s ()
+spendMemory term mem = do
+    spendSMemory term mem
+    spendWMemory term mem
+
+spendCPU :: WithMemory Term -> ExCPU -> CekM w s ()
+spendCPU term mem = do
+    spendSCPU term mem
+    spendWCPU term mem
+
+class Monoid w => SpendExUnitsWriter w where
+    spendWMemory :: forall s. WithMemory Term -> ExMemory -> CekMBase w s ()
+    spendWCPU :: forall s. WithMemory Term -> ExCPU -> CekMBase w s ()
+
+instance SpendExUnitsWriter () where
+    spendWMemory _ _ = pure ()
+    spendWCPU _ _ = pure ()
+
+class SpendExUnitsState s where
+    spendSMemory :: forall w. Monoid w => (WithMemory Term) -> ExMemory -> CekMBase w s ()
+    spendSCPU :: forall w. Monoid w => (WithMemory Term) -> ExCPU -> CekMBase w s ()
+
+instance SpendExUnitsState () where
+    spendSCPU _ _ = pure ()
+    spendSMemory _ _ = pure ()
+
+spendSBudget :: Ord a => Num a => Monoid w => Lens' ExBudget a -> WithMemory Term -> a -> CekMBase w ExBudget ()
+spendSBudget l term ex = do
+    newEx <- fmap (\x -> x - ex) $ gets (view l)
+    if newEx < 0 then
+        throwError $ MachineException OutOfExError (void term)
+    else
+        assign l newEx
```

Comment:
I'd be tempted to do:
```
do
    newEx <- l <-= ex
    when (newEx < 0) $ throw ...
```
This version does mean that the state is left with the negative exunits in it. But that might be useful so you can report the overrun?